crates/bevy_ecs/src/reflect/component.rs

```diff
@@ -273,12 +276,22 @@ impl<C: Component + Reflect + TypePath> FromType<C> for ReflectComponent {
                 entity.insert(component);
             },
             apply: |mut entity, reflected_component| {
-                let mut component = entity.get_mut::<C>().unwrap();
+                if !C::Mutability::MUTABLE {
+                    panic!("This component is immutable, you cannot modify it through reflection");
+                }
+
+                // SAFETY: guard ensures `C` is a mutable component
+                let mut component = unsafe { entity.get_mut_assume_mutable::<C>() }.unwrap();
                 component.apply(reflected_component);
             },
             apply_or_insert: |entity, reflected_component, registry| {
-                if let Some(mut component) = entity.get_mut::<C>() {
+                if !C::Mutability::MUTABLE {
+                    panic!("This component is immutable, you cannot modify it through reflection");
+                }
+
+                if let Some(mut component) = entity.take::<C>() {
                     component.apply(reflected_component.as_partial_reflect());
+                    entity.insert(component);
```

Comment:
Was this intentional?
```suggestion
```

Reply:
Simply removing the insert would be an error. The earlier version used `entity.get_mut()`, this one uses `entity.take()` so it needs to reinsert it back.

However, something is off here. Either the panic above is not needed, as mutating component through `take() + modify + insert()` is fully supported for immutable components, or this could still keep on using `entity.get_mut()` (unless there's just some reason why it doesn't compile).

What I believe instead is that `apply()` and `apply_or_insert()` should instead work on mutable components, and always do `take() + apply + insert()`. Or if we want to optimize for hypothetical performance, it can be an if that if the component is mutable, then it will be `get_mut() + apply`, otherwise through the take path.

`reflect_mut` should fail for immutable components obviously.

Reply:
Oh I totally glanced over the `take`! Then yeah the insert works.

Yeah I don't think we should be taking and reinserting. Wouldn't that unnecessarily trigger `OnAdd`/`Added`?

Either way, I do think we should keep the mutability check. Even if not technically necessary, it'll be way easier to remove than to add if we ever need to.

Reply:
I _think_ I wrote this before I added the `get_mut_assume_mutable` methods to the various `Entity` cells. You're right that it shouldn't `take`/`apply`/`insert` here, since we know via the `panic!` that at this point in the code-path we have a mutable component, making that option unnecessary.